### PR TITLE
[SPARK-16683][SQL] Repeated joins to same table can leak attributes via partitioning giving incorrect results

### PR DESCRIPTION
**Pull request by aray**
## What changes were proposed in this pull request?

In some complex queries where the same table is joined multiple times interleaved with aggregation we can get conflicting attributes that leak via partitionings leading to wrong results because shuffles are not inserted. See `JoinSuite` diff for example. This patch adds a method to `Partitioning` that restricts it to a given set of output attributes. This method is then called by operators that generally maintain their input distribution but output only a subset of the inputs.

## How was this patch tested?

Unit test based on example code from JIRA and additional unit testing of new method.

### CODE REVIEW

**Comment by aray** [comment]
Plan for the example query before the patch (with partitioning as suffix):
```
*HashAggregate(keys=[parent#228], functions=[], output=[level2#274]) hashpartitioning(parent#228, 5)
+- *HashAggregate(keys=[parent#228], functions=[], output=[parent#228]) hashpartitioning(parent#228, 5)
   +- *Project [parent#228] hashpartitioning(parent#228, 5)
      +- *BroadcastHashJoin [level1#270], [son#227], Inner, BuildRight hashpartitioning(parent#228, 5)
         :- *HashAggregate(keys=[parent#228], functions=[], output=[level1#270]) hashpartitioning(parent#228, 5)
         :  +- Exchange hashpartitioning(parent#228, 5) hashpartitioning(parent#228, 5)
         :     +- *HashAggregate(keys=[parent#228], functions=[], output=[parent#228]) UnknownPartitioning(0)
         :        +- *Project [parent#228] UnknownPartitioning(0)
         :           +- *BroadcastHashJoin [id#266], [son#227], Inner, BuildRight UnknownPartitioning(0)
         :              :- *Project [_1#264 AS id#266] UnknownPartitioning(0)
         :              :  +- *Filter isnotnull(_1#264) UnknownPartitioning(0)
         :              :     +- *SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple1, true])._1, true) AS _1#264] UnknownPartitioning(0)
         :              :        +- Scan ExternalRDDScan[obj#263] UnknownPartitioning(0)
         :              +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, string, false])) BroadcastPartitioning(HashedRelationBroadcastMode(List(input[0, string, false])))
         :                 +- *Filter (isnotnull(son#227) && isnotnull(parent#228)) UnknownPartitioning(0)
         :                    +- InMemoryTableScan [son#227, parent#228], [isnotnull(son#227), isnotnull(parent#228)] UnknownPartitioning(0)
         :                          +- InMemoryRelation [son#227, parent#228], true, 10000, StorageLevel(disk, memory, deserialized, 1 replicas), Statistics(sizeInBytes=148.0 B, hints=none)
         :                                +- *Project [_1#224 AS son#227, _2#225 AS parent#228] UnknownPartitioning(0)
         :                                   +- *SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._1, true) AS _1#224, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._2, true) AS _2#225] UnknownPartitioning(0)
         :                                      +- Scan ExternalRDDScan[obj#223] UnknownPartitioning(0)
         +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, string, false])) BroadcastPartitioning(HashedRelationBroadcastMode(List(input[0, string, false])))
            +- *Filter isnotnull(son#227) UnknownPartitioning(0)
               +- InMemoryTableScan [son#227, parent#228], [isnotnull(son#227)] UnknownPartitioning(0)
                     +- InMemoryRelation [son#227, parent#228], true, 10000, StorageLevel(disk, memory, deserialized, 1 replicas), Statistics(sizeInBytes=148.0 B, hints=none)
                           +- *Project [_1#224 AS son#227, _2#225 AS parent#228] UnknownPartitioning(0)
                              +- *SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._1, true) AS _1#224, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._2, true) AS _2#225] UnknownPartitioning(0)
                                 +- Scan ExternalRDDScan[obj#223] UnknownPartitioning(0)
```
and after the patch:
```
*HashAggregate(keys=[parent#228], functions=[], output=[level2#274]) UnknownPartitioning(5)
+- Exchange hashpartitioning(parent#228, 5) hashpartitioning(parent#228, 5)
   +- *HashAggregate(keys=[parent#228], functions=[], output=[parent#228]) UnknownPartitioning(5)
      +- *Project [parent#228] UnknownPartitioning(5)
         +- *BroadcastHashJoin [level1#270], [son#227], Inner, BuildRight UnknownPartitioning(5)
            :- *HashAggregate(keys=[parent#228], functions=[], output=[level1#270]) UnknownPartitioning(5)
            :  +- Exchange hashpartitioning(parent#228, 5) hashpartitioning(parent#228, 5)
            :     +- *HashAggregate(keys=[parent#228], functions=[], output=[parent#228]) UnknownPartitioning(0)
            :        +- *Project [parent#228] UnknownPartitioning(0)
            :           +- *BroadcastHashJoin [id#266], [son#227], Inner, BuildRight UnknownPartitioning(0)
            :              :- *Project [_1#264 AS id#266] UnknownPartitioning(0)
            :              :  +- *Filter isnotnull(_1#264) UnknownPartitioning(0)
            :              :     +- *SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple1, true])._1, true) AS _1#264] UnknownPartitioning(0)
            :              :        +- Scan ExternalRDDScan[obj#263] UnknownPartitioning(0)
            :              +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, string, false])) BroadcastPartitioning(HashedRelationBroadcastMode(List(input[0, string, false])))
            :                 +- *Filter (isnotnull(son#227) && isnotnull(parent#228)) UnknownPartitioning(0)
            :                    +- InMemoryTableScan [son#227, parent#228], [isnotnull(son#227), isnotnull(parent#228)] UnknownPartitioning(0)
            :                          +- InMemoryRelation [son#227, parent#228], true, 10000, StorageLevel(disk, memory, deserialized, 1 replicas), Statistics(sizeInBytes=148.0 B, hints=none)
            :                                +- *Project [_1#224 AS son#227, _2#225 AS parent#228] UnknownPartitioning(0)
            :                                   +- *SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._1, true) AS _1#224, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._2, true) AS _2#225] UnknownPartitioning(0)
            :                                      +- Scan ExternalRDDScan[obj#223] UnknownPartitioning(0)
            +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, string, false])) BroadcastPartitioning(HashedRelationBroadcastMode(List(input[0, string, false])))
               +- *Filter isnotnull(son#227) UnknownPartitioning(0)
                  +- InMemoryTableScan [son#227, parent#228], [isnotnull(son#227)] UnknownPartitioning(0)
                        +- InMemoryRelation [son#227, parent#228], true, 10000, StorageLevel(disk, memory, deserialized, 1 replicas), Statistics(sizeInBytes=148.0 B, hints=none)
                              +- *Project [_1#224 AS son#227, _2#225 AS parent#228] UnknownPartitioning(0)
                                 +- *SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._1, true) AS _1#224, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._2, true) AS _2#225] UnknownPartitioning(0)
                                    +- Scan ExternalRDDScan[obj#223] UnknownPartitioning(0)
```
Note there is now an `Exchange` inserted between the two top level `HashAggregate`

**Comment by aray** [comment]
retest this please

**Comment by aray** [comment]
ping @rxin can someone look at this correctness fix?

**Comment by rxin** [comment]
cc @cloud-fan @hvanhovell

**Comment by gatorsmile** [comment]
I will review this next week.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala [hunk 1]
```diff
@@ -80,7 +80,7 @@ case class ProjectExec(projectList: Seq[NamedExpression], child: SparkPlan)
 
   override def outputOrdering: Seq[SortOrder] = child.outputOrdering
 
-  override def outputPartitioning: Partitioning = child.outputPartitioning
+  override def outputPartitioning: Partitioning = child.outputPartitioning.restrict(outputSet)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Project won't change output partitioning, right?

**Comment by viirya** [comment]
I'd like to reword the problem description as the current one looks obscure to me.

Currently we don't care if the output partitioning of an operator contains the attributes not in the output. For example, the output partitioning of `Project [b, c]` can be hash partition of `[a]`. It is possible because it may come from:

    Project [b, c]
      Project [a, b, c] // output partitioning is hash partition of [a]
        Exchange hashpartitioning(a, 5)
          ...

I've noticed this and raised questions about it before. The answer I got is this doesn't do any harm so we don't fix it before.

However, this PR finds a case it possibly causes problem. Like:

    BroadcastHashJoin [a, b, c] // now the output partition is wrongly hash partition of [a]...
      Project [c] // a isn't in output, but we don't change output partitioning
        Project [a, b, c] // output partitioning is hash partition of [a]
          Exchange hashpartitioning(a, 5)
      BroadcastExchange [...]
        Project [a, b] // output partitioning is hash partition of [b]
          ...

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala [hunk 2]
```diff
@@ -65,6 +65,10 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
     false
   }
 
+  override def verboseStringWithSuffix: String = {
+    s"$verboseString $outputPartitioning"
+  }
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: Except for debugging this, do we really need to print out output partitioning always?

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 3]
```diff
@@ -36,6 +40,70 @@ class JoinSuite extends QueryTest with SharedSQLContext {
...
+      Tuple1("A21"),
+      Tuple1("B10"),
+      Tuple1("B11"),
+      Tuple1("B20"),
+      Tuple1("B21")
+    )).toDF("id")
+    base.createOrReplaceTempView("base")
+
+    val dist1 = spark.sql("""
+    SELECT parent level1
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: Please fix the code indent.

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 4]
```diff
@@ -36,6 +40,70 @@ class JoinSuite extends QueryTest with SharedSQLContext {
...
+    FROM base INNER JOIN hier h1 ON base.id = h1.son
+    GROUP BY parent""")
+
+    dist1.createOrReplaceTempView("dist1")
+    // dist1.count() // or put a count here
+
+    val dist2 = spark.sql("""
+    SELECT parent level2
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: ditto.

**Comment by viirya** [comment]
A different view to this problem is, in the following part of query plan:

         :- *HashAggregate(keys=[parent#228], functions=[], output=[level1#270]) hashpartitioning(parent#228, 5)
         :  +- Exchange hashpartitioning(parent#228, 5) hashpartitioning(parent#228, 5)
         :     +- *HashAggregate(keys=[parent#228], functions=[], output=[parent#228]) UnknownPartitioning(0)
         :        +- *Project [parent#228] UnknownPartitioning(0)

At the top `HashAggregate`, when the output is going to alias the attribute `parent` to `level1`, the output partitioning `hashpartitioning(parent, 5)` doesn't change accordingly to `hashpartitioning(level1, 5)`.

If we change `HashAggregate`'s output partitioning accordingly with its output, this query can return correct results.

I think this is also an alternative solution. @aray What do you think?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala (hunk 2 above)
Comment (aray) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: This doesn't change anything that is in common use, one has to do `plan.treeString(verbose = true, addSuffix = true)` to get it. I would argue for keeping it for any future debugging.

**Comment by aray** [comment]
@viirya We could certainly make that improvement. I believe it would be a fairly trivial change to this PR if we were just considering expressions that have the same canonical representation. However for reasons that are not clear to me an alias does not automatically have the same canonical representation as the `exprId` is not copied (https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Column.scala#L984). Can anyone enlighten me as to why this is the case?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 5]
```diff
@@ -171,6 +171,16 @@ sealed trait Partitioning {
    * produced by `A` could have also been produced by `B`.
    */
   def guarantees(other: Partitioning): Boolean = this == other
+
+  /**
+   * Returns the partitioning scheme that is valid under restriction to a given set of output
+   * attributes. If the partitioning is an [[Expression]] then the attributes that it depends on
+   * must be in the outputSet otherwise the attribute leaks.
+   */
+  def restrict(outputSet: AttributeSet): Partitioning = this match {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: We are refactoring the concept in the PR https://github.com/apache/spark/pull/19080

Please review that PR first? Thanks!

**Comment by cloud-fan** [comment]
shouldn't we fix `ProjectExec.outputPartitioning`?

**Comment by viirya** [comment]
If we have correct `outputPartitioning` on each operators, I think this issue can be fixed.

**Comment by maropu** [comment]
@aray Can you close this for now because it's not active for a long time? (I'm not sure the current master still has this issue..., so you should check it first)

**Comment by HyukjinKwon** [comment]
Let't close this then.

**Pull request closed without merging**